ragna/source_storages/_qdrant.py

```diff
@@ -0,0 +1,249 @@
+from __future__ import annotations
+
+import os
+import uuid
+from collections import defaultdict
+from typing import TYPE_CHECKING, Any, Optional, cast
+
+import ragna
+from ragna.core import (
+    Document,
+    MetadataFilter,
+    MetadataOperator,
+    PackageRequirement,
+    Requirement,
+    Source,
+)
+
+from ._utils import raise_no_corpuses_available, raise_non_existing_corpus
+from ._vector_database import VectorDatabaseSourceStorage
+
+if TYPE_CHECKING:
+    from qdrant_client import models
+
+
+class QdrantDB(VectorDatabaseSourceStorage):
+    """[Qdrant vector database](https://qdrant.tech/)
+
+    !!! info "Required packages"
+
+        - `qdrant-client>=1.12.0`
+    """
+
+    DOC_CONTENT_KEY = "__document"
+
+    @classmethod
+    def requirements(cls) -> list[Requirement]:
+        return [
+            *super().requirements(),
+            PackageRequirement("qdrant-client>=1.12.1"),
+        ]
+
+    def __init__(self) -> None:
+        super().__init__()
+
+        from qdrant_client import QdrantClient
+
+        url = os.getenv("QDRANT_URL")
+        api_key = os.getenv("QDRANT_API_KEY")
+        path = path = ragna.local_root() / "qdrant"
+
+        # Cannot pass both url and path
+        self._client = (
+            QdrantClient(url=url, api_key=api_key) if url else QdrantClient(path=path)
+        )
```

Comment:
IIUC, you only need the API key if you are using a remote database, i.e. one that needs an URL to connect to it. Will the `QdrantClient` raise an informative error message when the API key is not set?

Regardless, I'd like to refactor this slightly to make it more obvious what is happening

```suggestion
        
        if (url := os.getenv("QDRANT_URL")) is not None:
            kwargs = dict(
                url=url,
                api_key = os.getenv("QDRANT_API_KEY"),
            )
        else:
            kwargs = dict(ragna.local_root() / "qdrant")

        self._client = QdrantClient(**kwargs)
```

Reply:
Not only cloud, but a local server with Docker.
```
docker run -p 6333:6333 qdrant/qdrant
```
URL would be http://localhost:6333.

Note that, when you pass the local file `path` argument instead of the `url`, the client does a Python based imitation of the Qdrant server APIs. This should only be used for prototyping and CI.

For real use cases, Qdrant(Rust based) server should be used.

Reply:
The current snippet should be fine IMO.

Reply:
In that case only minor cleanup :slightly_smiling_face: 

```suggestion
        url = os.getenv("QDRANT_URL")
        api_key = os.getenv("QDRANT_API_KEY")
        path = ragna.local_root() / "qdrant"

        # Cannot pass both url and path
        self._client = (
            QdrantClient(url=url, api_key=api_key) if url else QdrantClient(path=path)
        )
```

Reply:
Great save! Thanks.